### PR TITLE
Fix: initialization fails with message "waitpid failed for mtd"

### PR DESCRIPTION
**Pull request by mstrinzha**

### CODE REVIEW

File: libraries/AP_BoardConfig/px4_drivers.cpp [hunk 1]
```diff
@@ -267,7 +267,26 @@ bool AP_BoardConfig::px4_start_driver(main_fn_t main_function, const char *name,
...
+    if (ret < 0)
+    {
+      int errcode = errno;
+      /* Unfortunately, this main thread does not retain child status.  If
+       * child status is enabled (via CONFIG_SCHED_CHILD_STATUS), ostest_main()
+       * disables the feature by calling sigactin with SA_NOCLDWAIT.
+       */
+      if (errcode == ECHILD)
+      {
+          free(s);
```
Comment (EShamaev) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: Can free(s); be moved outside of if clause?

Comment (magicrub) [inline libraries/AP_BoardConfig/px4_drivers.cpp]: Yeah, could be
Free
Return (errcode !=Echild)

**Comment by mstrinzha** [comment]
Code is updated.

**Comment by tridge** [comment]
how do we reproduce the problem that this fixes?
If it is giving that error message then that means mtd is not starting correctly.

**Comment by mstrinzha** [comment]
https://github.com/Swift-Flyer/ardupilot/blob/0555fb03400c0ed7293e15377e4ab08b863462a5/libraries/AP_BoardConfig/px4_drivers.cpp#L261
main_function is finished before waitpid is called. waitpid fails.

**Comment by mstrinzha** [comment]
synthetic test
int test_fast_main_function(int, char **) {
  return 0; //nothing to do
}

px4_start_driver(test_fast_main_function, "test_function", null);

**Comment by OXINARF** [comment]
Interesting, apparently NuttX doesn't keep child status after death by default. We can turn it on (https://github.com/ArduPilot/PX4NuttX/blob/master/nuttx/sched/Kconfig#L80-L119) but may be more interesting to us to just use the workaround specified there.

**Comment by tridge** [comment]
my only concern is this may hide the failure of one of the drivers we depend on.
Which driver are you trying to start that is failing?
When I first added this function I had to modify some of the main functions to behave correctly.
One solution is to add a boolean to the function:

```
bool AP_BoardConfig::px4_start_driver(main_fn_t main_function, const char *name, 
                                  const char *arguments, bool check_return=true)


```

**Comment by tridge** [comment]
actually, the suggestion to use sched_lock() and sched_unlock() in the link from @OXINARF looks good

**Comment by khancyr** [comment]
Closed as we don't support PX4 anymore

**Pull request closed without merging**